src/services/text/TextGenerationService.ts

```diff
@@ -10,70 +9,90 @@ import {
   TextGenerationCreateStreamOutput,
 } from '../../schema.js';
 import { TypedReadable } from '../../utils/stream.js';
+import { ApiClient } from '../../api/client.js';
+import { SteamingApiClient } from '../../api/streaming-client.js';
+import { ConcurrencyLimiter } from '../../utils/concurrency.js';
+import { BaseService } from '../BaseService.js';
 
 export class TextGenerationService extends BaseService {
-  create(
+  constructor(
+    protected readonly _client: ApiClient,
+    protected readonly _streamingClient: SteamingApiClient,
+    protected readonly _limiter: ConcurrencyLimiter,
+  ) {
+    super(_client, _streamingClient);
+  }
+
+  async create(
     input: TextGenerationCreateInput,
     opts?: Options,
   ): Promise<TextGenerationCreateOutput> {
-    return clientErrorWrapper(
-      this._client.POST('/v2/text/generation', {
-        ...opts,
-        params: { query: { version: '2024-01-10' } },
-        body: input,
-      }),
+    return this._limiter.execute(
+      () =>
+        clientErrorWrapper(
+          this._client.POST('/v2/text/generation', {
+            ...opts,
+            params: { query: { version: '2024-01-10' } },
+            body: input,
+          }),
+        ),
+      { signal: opts?.signal },
     );
   }
 
   create_stream(
     input: TextGenerationCreateStreamInput,
     opts?: Options,
-  ): TypedReadable<TextGenerationCreateStreamOutput> {
+  ): Promise<TypedReadable<TextGenerationCreateStreamOutput>> {
     type EventMessage = TextGenerationCreateStreamOutput;
+    return this._limiter.execute(
+      async () => {
+        const stream = new Transform({
+          autoDestroy: true,
+          objectMode: true,
+          transform(
+            chunk: EventMessage,
+            encoding: BufferEncoding,
+            callback: TransformCallback,
+          ) {
+            try {
+              const {
+                generated_text = '',
+                stop_reason = null,
+                input_token_count = 0,
+                generated_token_count = 0,
+                ...props
+              } = (chunk.results || [{}])[0];
 
-    const stream = new Transform({
-      autoDestroy: true,
-      objectMode: true,
-      transform(
-        chunk: EventMessage,
-        encoding: BufferEncoding,
-        callback: TransformCallback,
-      ) {
-        try {
-          const {
-            generated_text = '',
-            stop_reason = null,
-            input_token_count = 0,
-            generated_token_count = 0,
-            ...props
-          } = (chunk.results || [{}])[0];
+              callback(null, {
+                generated_text,
+                stop_reason,
+                input_token_count,
+                generated_token_count,
+                ...(chunk.moderation && {
+                  moderation: chunk.moderation,
+                }),
+                ...props,
+              });
```

Comment:
This should not be there anymore. We should stream the whole chunk and not do any post-processing.